ir/memory.cpp

```diff
@@ -431,6 +431,39 @@ Memory::Memory(State &state, bool little_endian)
   local_block_liveness
     = expr::mkLambda({ expr::mkVar("#bid0", bitsBid() - 1) }, false);
 
+  // The first byte of the memory cannot be allocated because it is for a null
+  // pointer.
+  // The last byte of the memory cannot be allocated because ptr + size
+  // (where size is the size of the block and ptr is the beginning address of
+  // the block) should not overflow.
+  avail_space = expr::mkVar("avail_space", bitsPtrSize());
+  if (state.isSource()) {
+    state.addAxiom(avail_space.ult(expr::mkUInt(-2, bitsPtrSize())));
+  }
+
+  // TODO: replace the magic number 2 with the result of analysis.
+  // This is just set as 2 to make existing unit tests run without timeout.
+  unsigned non_local_bid_upperbound = 2;
+  // Maximum alignment that a non-local block can have
+  // TODO: get this from data layout?
+  auto max_align_minus1 = expr::mkUInt(8 - 1, bitsPtrSize());
+
+  // Omit null pointer
+  for (unsigned bid = 1; bid <= non_local_bid_upperbound; ++bid) {
+    Pointer p(*this, bid, false);
+
+    // The required space size should consider alignment
+    auto size_upperbound = p.block_size() + max_align_minus1;
+
+    if (state.isSource()) {
+      state.addAxiom(p.is_block_alive().implies(
+          size_upperbound.ule(avail_space)));
+    }
+    avail_space = avail_space - expr::mkIf(p.is_block_alive(),
+                                           p.block_size(),
```

Comment:
this should use size_upperbound instead

Reply:
I think block_size is okay here, because smaller blocks can fit into the empty space which was forced by alignment.

Reply:
If you need to align this allocation, then you effectively waste the alignment space.
Otherwise this encoding may decide there's still space when there's not. For example:
p = alloc 4
q = alloc 4
r = alloc 4

Initial space: 12. Align to 8.
allocations succeed, because there's enough space. But disjointness constraint fails:
we have:
p = 0
q = 8
r = 12 ?? impossible -> UNSAT

Reply:
The case is addressed by line 536 (`allocated &= size_upperbound.ule(avail_space);`); the third alloca isn’t allocated because size_upperbound is larger than avail_space.

Using size addresses this case:
P = alloc 4, align 4
Q = alloc 2, align 2
Assuming that initial space is 8, it is possible that p = 0x04 & q = 0x06 , but subtracting size_upperbound would block this.

Reply:
It is line 458, not 536 (`state.addAxiom(p.is_block_alive().implies(size_upperbound.ule(avail_space)))`), but rest of the comment is still valid I think.

Reply:
If you don't update the available size with the alignment size, you don't get the compounding effect of multiple alignments. The side constraints you have aren't sufficient because they only check that the summation of the sizes + the last alignment fit.

Reply:
Understood, so to correctly simulate fragmentation, size_upperbound should be added.